### DISCUSSION TITLE
Dynamic commands

### DISCUSSION THREAD

**Discussion by aminchegeni** [spring-projects/spring-shell#965] [Ideas]
Hi Mr. Valkealahti

Happy new Years

I search for the document for in-depth knowledges about internal and design of spring shell. I read a lot official documents and did many debug around internals of spring shell. I decide to design cli application that knows its commands at runtime. It connects to the some rest endpoint and detects the actual commands can be provided. In my design I like to add some features as below:

•	Enable some aliasing like Linux (for example the user can alias some commands by its default as new command to CommandCatalog)
•	Add CommandRegistration instance that has many options but Its TargetSpec bound to the bean’s method with single DTO (in fact the command options are fields of target method input DTO. I implemented it by using the HandlerMethodArgumentResolver).
•	I need mechanism for add new instance of HandlerMethodArgumentResolver to CommandExecutionHandlerMethodArgumentResolvers (currently I used @EventListener for ContextRefreshedEvent and customize default ones)
•	I need know about some internal part of spring shell

I really will be happy if you could help me by some hints or advanced docs.

Best wishes for you in the new year

Amin Chegeni

**Comment by jvalkeal** [comment]
Defined commands can already define aliases but looks like you'd want to have a feature where _user_ can define these at runtime. That we don't have but it's a relatively nice idea. You could create a new issue for this.

For dynamic commands there's a `CommandResolver` which can be used to add commands which are not known at build time. There's an example of its use in `spring-cli` [DynamicMethodCommandResolver](https://github.com/spring-projects/spring-cli/blob/main/src/main/java/org/springframework/cli/runtime/command/DynamicMethodCommandResolver.java) Is this something you're after?

**Reply by aminchegeni** [reply]
Hi Mr. Valkealahti

Thank you for your reply

Please let me check spring-cli asap.

I will come back here for additional explanation.

Best regards
